[ONNX] Supports scatter_add with different static shape of src and index

titaiwangms opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #89787


Prior to this change, the converter doesn't support `scatter_add` with different shape of `src` and `index`, while [it's claimed to be supported by PyTorch](https://pytorch.org/docs/stable/generated/torch.Tensor.scatter_add_.html#torch.Tensor.scatter_add_) in a way that scatter shape would be accommodated to index shape. This PR adds `onnx::Slice` to adjust the shape of `src` when a static and mismatched shape is found. However, if both of the shape (src and index) is set to dynamic, they are expected to be the same shape from ONNX due to the spec. More ScatterElements details on https://github.com/onnx/onnx/issues/4672

titaiwangms committed:
[ONNX] Supports scatter_add with different static shape of src and index

[ghstack-poisoned]

BowenBao commented on torch/onnx/symbolic_opset16.py:
```diff
@@ -75,12 +75,47 @@ def scatter_add(g: jit_utils.GraphContext, self, dim, index, src):
+                    "scatter_add",
+                    f"`index` ({index_sizes}) should have smaller or equal (<=) size at any dimension than `src` ({src_sizes})",
+                )
+            axes.append(idx)
```
Should this happen only when `src_sizes[idx] > d`? No need to slice when they are equal right?

titaiwangms replied:
Done

BowenBao commented on torch/onnx/symbolic_opset16.py:
```diff
@@ -75,12 +75,47 @@ def scatter_add(g: jit_utils.GraphContext, self, dim, index, src):
+                g, src, axes=axes, starts=starts, ends=ends
+            )
+        # Align the dynamic sizes of src and index
+        # NOTE: Even if users set src and index with different dynamic axes, they are
```
It looks like symbolic function does not have enough information to remediate if shape don't match under dynamic size. 
Does setting sizes manually below has any effect? (Is it to pass onnx checker?) If not I'd suggest remove it and comment above why mismatch under dynamic shape cannot be fixed here.

titaiwangms replied:
Done

BowenBao reviewed (approved):
LGTM, a few suggestions.

titaiwangms committed:
Update on "[ONNX] Supports scatter_add with different static shape of src and index"



Prior to this change, the converter doesn't support `scatter_add` with different shape of `src` and `index`, while [it's claimed to be supported by PyTorch](https://pytorch.org/docs/stable/generated/torch.Tensor.scatter_add_.html#torch.Tensor.scatter_add_) in a way that scatter shape would be accommodated to index shape. This PR adds `onnx::Slice` to adjust the shape of `src` when a static and mismatched shape is found. However, if both of the shape (src and index) is set to dynamic, they are expected to be the same shape from ONNX due to the spec. More ScatterElements details on https://github.com/onnx/onnx/issues/4672

[ghstack-poisoned]

titaiwangms commented:
@pytorchbot merge -g

titaiwangms commented:
@pytorchbot merge -g
